adapters/philips-hue/philips-hue-adapter.js

```diff
@@ -0,0 +1,233 @@
+   * @param {Object} light - the light API object
+   */
+  constructor(adapter, id, lightId, light) {
+    super(adapter, id);
+
+    this.lightId = lightId;
+    this.name = light.name;
+
+    this.type = THING_TYPE_ON_OFF_SWITCH;
+    this.properties.set('on', new PhilipsHueProperty(this, 'on', 'boolean',
+                        light.state.on));
+
+    this.adapter.handleDeviceAdded(this);
+  }
+
+  /**
+   * When a property changes notify the Adapter to communicate with the bridge
+   * @param {PhilipsHueProperty} property
+   */
+  notifyPropertyChanged(property) {
+    super.notifyPropertyChanged(property);
+    var on = this.properties.get('on').value;
+    this.adapter.sendProperties(this.lightId, {on: on});
+  }
+}
+
+/**
+ * Philips Hue Bridge Adapter
+ * Instantiates one PhilipsHueDevice per light
+ * Handles the username acquisition (pairing) process
+ */
+class PhilipsHueAdapter extends Adapter {
+  constructor(adapterManager, bridge) {
+    super(adapterManager, 'philips-hue');
+
+    this.username = null;
+    this.bridge = bridge;
+    this.bridgeIp = bridge.internalipaddress;
+    this.pairing = false;
+    this.lights = {};
+
+    adapterManager.addAdapter(this);
+
+    storage.init().then(() => {
+      return storage.getItem(KNOWN_BRIDGE_USERNAMES);
+    }).then(knownBridgeUsernames => {
+      if (!knownBridgeUsernames) {
+        return Promise.reject('no known bridges');
+      }
+
+      var username = knownBridgeUsernames[this.bridge.id];
+      if (!username) {
+        return Promise.reject('no known username');
+      }
+      this.username = username;
+      this.discoverLights();
+    }).catch(e => {
+      console.error('philips-hue-adapter storage error:', e);
+    });
+  }
+
+  /**
+   * If we don't have a username try to acquire one from the bridge
+   * @param {number} timeoutSeconds
+   */
+  startPairing(timeoutSeconds) {
+    this.pairing = true;
+    var endTime = Date.now() + timeoutSeconds * 1000;
+
+    var attemptPairing = () => {
+      this.pair().then(username => {
+        this.username = username;
+        return this.discoverLights();
+      }).then(() => {
+        return storage.init();
+      }).then(() => {
+        return storage.getItem(KNOWN_BRIDGE_USERNAMES);
+      }).then(knownBridgeUsernames => {
+        if (!knownBridgeUsernames) {
+          knownBridgeUsernames = {};
+        }
+        knownBridgeUsernames[this.bridge.id] = this.username;
+        return storage.setItem(KNOWN_BRIDGE_USERNAMES, knownBridgeUsernames);
+      }).catch(e => {
+        console.error('philips-hue-adapter pairing error:', e);
+        if (this.pairing && Date.now() < endTime) {
+          // Attempt pairing again later
+          setTimeout(attemptPairing, 500);
+        }
+      });
+    };
+
+    attemptPairing();
+  }
+
+  /**
+   * Perform a single attempt at pairing
+   * @return {Promise} Resolved with username if pairing succeeds
```

Comment:
Pairing a hub then discovering lights, I updated the comment to be a bit more clear